### ISSUE TITLE
Restoring positions after redocking does not work

### ISSUE THREAD

**Issue by aronovgj** [manutalcual/winredock#10]
- After redocking window positions and sizes are restored to one monitor
- All other monitors are empty
- Reading the config again does not fix it

**Comment by manutalcual** [comment]
Reading the config will not solve anything now. The app is now different. But I have tested it and it is very strange that it doesn't works for you.
Could you be so gentle to run a special logging version to see your screen sizes and what the app is receiving in terms of internal Windows messages?

I don't have any Windows machine at home (I'm a Linux guy) and have to do this at work, but obviously I'm limited to my laptop and screens, and  it works on mine so... I can't do more without help.

Thank you for your time and help!

**Comment by aronovgj** [comment]
Sure, just give me the link to the logging version.

**Comment by manutalcual** [comment]
I've created a branch to solve this issue: I0010-restoring-positions-doesnt-work
You can find the logging binary there: [bin](https://github.com/manutalcual/winredock/blob/I0010-restoring-positions-doesnt-work/bin/bin_logging.zip)

**Comment by aronovgj** [comment]
[wm.log](https://github.com/manutalcual/winredock/files/2610578/wm.log)

**Comment by manutalcual** [comment]
Well... I understand why it doesn't works for you. When you undock the screen size doesn't changes. I don't know why.

This is before undocking, 1920x1080

    [../../src/window.hh:261] Receive WM_TIMER event.
    [../../src/poshandler.cc:92] Getting current desktop windows.
    [../../src/dev.hh:51] [dev] Screen: 1920, 1080

After undocking:

```
[../../src/window.hh:273] WM_DEVICECHAGE received!!!!
[../../src/window.hh:286] Message: DBT_DEVNODES_CHANGED
[../../src/window.hh:288] Screen size...
[../../src/dev.hh:51] [dev] Screen: 1920, 1080
[../../src/window.hh:290] New screen...
[../../src/dev.hh:51] [dev] Screen: 1920, 1080
[../../src/window.hh:292] Last screen...
[../../src/dev.hh:51] [dev] Screen: 1920, 1080

```
As you may see there is no resolution change: always 1920x1080.
I'm going to investigate this on Google to see if there is someone who has find this very problem before and why is it caused.

**Comment by aronovgj** [comment]
Oh, this might me connected:

The screens are connected via a docking station.
The screen settings in Windows show all three screens as connected but when I open Intels graphics settings I only see one screen. 
If I connect the screens directly with the laptop, without a docking station, I can see all screens in the windows settings and in the intel settings.

So you might need another way to access information about the screens.

**Comment by manutalcual** [comment]
Ok. I've found related windows info:

> GetSystemMetrics(SM_CMONITORS) counts only visible display monitors. This is different from EnumDisplayMonitors, which enumerates both visible display monitors and invisible pseudo-monitors that are associated with mirroring drivers. An invisible pseudo-monitor is associated with a pseudo-device used to mirror application drawing for remoting or other purposes.

So I have to EnumDisplayMonitors and act upon that information.
Let me some time (hopefully not much) to generate a new version.

Meanwhile, could you try the new version connecting the screens without the docking station? (Just to confirm the problem, *not* to take as a solution.)

**Comment by aronovgj** [comment]
Two tries with only one monitor connected directly to laptop:

[wm.log](https://github.com/manutalcual/winredock/files/2610680/wm.log)
[wm1.log](https://github.com/manutalcual/winredock/files/2610681/wm1.log)

Sadly, did not work either...

**Comment by manutalcual** [comment]
Ok, I've managed to reproduce the failure. I'm going to work on this. It may take some time as I can't work on this just now.

**Comment by manutalcual** [comment]
I've updated the loggin zip. Now it works in my laptop when:

- undocking, re-docking
- disabling one monitor through Windows "display settings"

Hope it works for you now. Tell me if it is OK or not.

**Comment by aronovgj** [comment]
Nope, sorry...

After undocking and redocking 2 monitors, still same behavior:

[wm.log](https://github.com/manutalcual/winredock/files/2611087/wm.log)

**Comment by manutalcual** [comment]
I'm going to take the laptop at home and work on thins over the weekend.

**Comment by manutalcual** [comment]
Another logging version uploaded. There should be no behavior changes, just more log. I've seen there are two way to obtain the screen/s resolution. Let's check the second one.

**Comment by aronovgj** [comment]
This time most windows were restored (it took more time than in previous versions). Some windows were moved to other desktops and were not maximized. But they did not all stay on the same screen this time after redocking.
[wm.log](https://github.com/manutalcual/winredock/files/2623414/wm.log)

Could you please anonymize window names and app names in the next version? It takes quite some time to go through the logs and fix this.

**Comment by aronovgj** [comment]
[wm.log](https://github.com/manutalcual/winredock/files/2623431/wm.log)

**Comment by aronovgj** [comment]
[wm.log](https://github.com/manutalcual/winredock/files/2623471/wm.log)

On second try it did not work again...

**Comment by manutalcual** [comment]
Yes. Sorry for the anonymization. I should have thought about that before asking for help. Sorry.

Looking into the logs, it seems the resolution is: width: 5760, height: 1080.
So the same resolution for every monitor: 1920x1080, but negative axis to position in different monitors: 
```
[../../src/poshandler.cc:74] Adding window with class 'rctrl_renwnd32': , top 0, left -1548, right -74, bottom 862
```
Plus, there is a moment when there is only 3 monitors attached, and it reports 4:

```
[../../src/window.hh:262] Receive WM_TIMER event.
[../../src/dev.cc:7] get_ennum_windows called
[../../src/dev.cc:7] get_ennum_windows called
[../../src/dev.cc:7] get_ennum_windows called
[../../src/dev.hh:64] Metrics: width: 5760, height: 1080, right: 1920, bottom: 1080, monitors: 4
[../../src/dev.hh:70] [dev] Screen: 1920, 1080 (mon 4)
```
When screen size increments (3mon. => 4) I get the new positions. This fourth monitor must be a _virtual_ one to draw or prepare to draw windows on it. Then, when I get the positions they are already for one monitor:
```
[../../src/poshandler.cc:74] Adding window with class 'rctrl_renwnd32': , top 0, left 434, right 1908, bottom 862
```
As you can see, the position isn't negative anymore. It was left -1548, right -74.
So either Windows or the device driver is drawing window contents in a hidden memory buffer and after the draw, its bits are overwritten in the graphical card memory (or something in those lines.)

So I have to ignore the positions when... the system resolution is the same as the max resolution detected until now (5760x1080) and there are more monitors than there were when that resolution was first detected. Clearer: if I have 5760 with two monitors, I have to ignore positions when I have 5760 with three or more monitors.

Unfortunately I don't have a system doing those things with screen resolution and positioning, so I'm going to try several algorithms that may be activated by an app parameter to see if I can fix it.

**Comment by manutalcual** [comment]
I've uploaded a new version. It now takes in account not only the monitor count but also the _virtual_ screen size (with all monitors.) I think this should work, so no parameters and one algorithm only.
Cross your fingers!
(Anonymized!)

**Comment by aronovgj** [comment]
Also, to make sure that we are on the same page: I talk about Desktops and Screens. 

- By Screens I mean the actual monitor 
- By Desktops I mean the Windows 10 feature that they stole from Linux where you can have several desktops and switch between them via shortcuts so you effectively have for example `3 monitors x 5 desktops = same space like having 15 monitors`.

And there are two issues: 
1. Windows are restored to one screen
2. Some windows change the desktop while being restored

We are currently trying to fix issue 1.
https://github.com/manutalcual/winredock/issues/9 is issue 2.

**Comment by aronovgj** [comment]
Some windows were restored, some were not. Some windows changed desktops.

[wm.log](https://github.com/manutalcual/winredock/files/2657222/wm.log)

**Comment by manutalcual** [comment]
Another try to solve the issue. It works like a charm for me.

**Comment by manutalcual** [comment]
Hey, welcome back!
I'm going to investigate.

**Comment by ReneCD** [comment]
Just fell over this handy utility after I for god-knows-what time redocked my Win10-Thinkpad-Onelink+ and had to re-sort all my windows.  Unfortunately it doesn't seem to work for me.  I downloaded the bin, extracted and ran. Tried clicking around with the get/save/read buttons both before and after redocking (several times) but nothing seems to happen. Don't know if I am missing a crucial step?  Pity though, it's a good idea!!

**Comment by manutalcual** [comment]
Hi,

Try branch restoring-positions-doesnt-works or something like that. In that branch you don't need to use the menu, all is done automatically (when it works.)
Unfortunately, Microsoft apps are more or less redundant to be handled by others, so they usually don't work.
And as a shame, I don't have enough time to investigate or develop further this tool. It works for me but it doesn't give anything else to me and I have to pay the bills. Sorry.

**Comment by ReneCD** [comment]
Fair enough, I can understand that. No worries, I'll keep looking.

**Comment by hansharhoff** [comment]
The version in the I0010-restoring-positions-doesnt-work branch works for me. Perhaps this should be merged to master so that this is the version people find by default?

**Comment by manutalcual** [comment]
Yes. Absolutely.

**Comment by manutalcual** [comment]
Done.

**Comment by manutalcual** [comment]
And I'm glad it works for you man! I'm quite happy with it even when some windows (mostly MS) refuse to play fair.

**Comment by hansharhoff** [comment]
Thank you!
I just started using it, so I am crossing my fingers.

/Hans
